doc/adr/0004-marconi-initiative.rst

```diff
@@ -0,0 +1,100 @@
+
+This architectural decision has some desirable effects:
+
+  1. Managing rollbacks is very simple and fast. We drop the events that were rolled back. (no complicated logic is required to undo the projection of the list of events on disk, which we would need if we would want to store everything on disk as fast as possible). 
+  
+  2. Making 'K' configurable makes the design already quite scalable. Developers do not usually need to guard themselves against rollbacks by K blocks so they can choose to only store 10 events in memory thus sacrificing chain integrity for RAM.
+  
+  3. In case of a restart recovery is very simple. We only store fully confirmed transactions so there is nothing to do other than resume operation.
+
+And some less desirable effects:
+
+  1. We must keep K events in memory, which (depending on how large events are) can waste some memory. Our educated guess is that this is a reasonable compromise, but depending on how large events can get that may not be the case for your use case.
+  
+  2. Queries are more complicated as we need to query both events in memory and the state persisted on disk.
+
+=================
+Query and storage
+=================
+
+The indexed data is accessible through queries. There are no constraints on the format of queries or results. Both are identified by a type variable that the indexer exposes and the code for both of them has to be provided by the user.
+
+This is currently the only area where the users can customise the indexer by providing functions that handle queries and storage. Note that the queries have to run on both the in-memory data and the data that is stored.
+
+This allows us to associate any kind of storage type to the indexers, though, right now we are only using SQLite.
+
+==================
+Versioned contents
+==================
+
+We need some sort of versioning or sequencing of events, notifications and queries.
+
+  1. Synchronisation of multiple indexers (queries have a validity interval)
+  2. Resume functionality (we need to know from which slot to resume)
+  3. Handling of rollbacks (now there is explicit handling of rollbacks)
+
+More information will become available in the next few sprints.
+
+=============
+Event streams
+=============
+
+To support PAB functionality which subscribes to a source for a set of event types we need a way to produce events from indexers.
+
+They are also very useful for contracts that want to track rollbacks. Rollbacks are invisible from the point of view of the indexed data, but it may be the case that the internal state of a contract needs to know that the state has been reverted.
+
+============
+Similar work
+============
+
+One of the advantages that Marconi has over Oura and Scrolls is the fact that they are both a streaming solution and an indexer. So you can listen to a stream of events and know that those events are reflected in the index.
```

Comment:
I'm slightly unclear on the meaning. Can you clarify a little bit? Is this saying that Marconi is both a streaming solution and an indexer? I'm not sure what the word "they" is referring to --> "they are both a streaming solution and an indexer" 

I like this document! Very informative.

Reply:
I need to rewrite this, but I think I need a little bit of time to look over the competing solutions to make sure I am not assuming too many things about them.